### PR TITLE
1D and 2D energy spectra for GCM

### PR DESCRIPTION
**Pull request by LenkaNovak**
# Description

Co-authored by @jiahe23 , @kpamnany 

This code calculates 1d spectra (Fourier transform at each latitude and height) and 2d spectra (spherical harmonic transformation at each height) for the GCM on the fly, or it can be ported to analysis scripts for offline calculations. 

It is currently set up to be interpolated using the same output timestep and Gaussian grid as the `atmos_default_diagnostics` group, but these can be set separately (e.g. for less severe wavenumber truncations). 

Energy spectra are currently calculated using the meridional wind only.

Output is a separate NetCDF file, containing:
"spectrum_1d"[zonal wavenumber, latitude, height, time]
"spectrum_2d"[truncated zonal wavenumber, total wavenumber, height, time]
 
TODO in a separate PR
- add mass weighting and vertical integration

_The code structure follows Yassine’s 3d turbulence LES spectra, and the spherical harmonic transformation is based on Daniel’s Julian FMS code._

I have

- [x] Written and run all necessary tests with CLIMA by including `tests/runtests.jl`
- [x] Followed all necessary [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and run `julia .dev/climaformat.jl .`
- [x] Updated the documentation to reflect changes from this PR.

# For review by CLIMA developers

- [ ] There are no open pull requests for this already
- [ ] CLIMA developers with relevant expertise have been assigned to review this submission
- [ ] The code conforms to the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) and has consistent naming conventions. `julia .dev/format.jl` has been run in a separate commit.
- [ ] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

File: src/Common/Spectra/spherical_helper.jl [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+  wi = 2/(1-xi^2)/P_n'(xi)^2
+
+# Arguments
+- n: number of latitudes
+
+# References
+- Ehrendorfer, M., Spectral Numerical Weather Prediction Models, Appendix B, Society for Industrial and Applied Mathematics, 2011
+
+"""
+function ComputeGaussian!(n)
```
Comment (kpamnany) [inline src/Common/Spectra/spherical_helper.jl:124]: See the [coding style](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) guidelines -- function names  should be lowercase with words separated by underscores as necessary to improve readability.

File: src/Diagnostics/Diagnostics.jl [hunk 2]
```diff
@@ -14,7 +14,9 @@ export DiagnosticsGroup,
     setup_atmos_mass_energy_loss,
     setup_dump_state_diagnostics,
     setup_dump_aux_diagnostics,
-    setup_dump_spectra_diagnostics
+    setup_dump_spectra_diagnostics,
+    setup_dump_gcm_spectra_diagnostics
```
Comment (kpamnany) [inline src/Diagnostics/Diagnostics.jl:18]: This should be added in `setup_dump_spectra_diagnostics` itself. Will discuss how on Slack.

File: src/Driver/diagnostics_configs.jl [hunk 3]
```diff
@@ -29,12 +29,14 @@ Creates an `InterpolationTopology` (either an `InterpolationBrick` or an
 function InterpolationConfiguration(
     driver_config::DriverConfiguration,
     boundaries::Array,
-    resolution::Tuple,
+    resolution::Tuple;
+    lat_range = nothing,
+    lon_range = nothing,
```
Comment (kpamnany) [inline src/Driver/diagnostics_configs.jl:34]: We should probably rework this -- add a new constructor. Again, will discuss.

**Review by kpamnany — CHANGES_REQUESTED** [review]

File: src/Common/Spectra/spherical_helper.jl (hunk 1 above)
Comment (LenkaNovak) [reply src/Common/Spectra/spherical_helper.jl:124]: Yep, I was planning to address the formatting once the functionality is tested by Jia and Daniel, but thanks for flagging this up! :)

File: experiments/AtmosGCM/GCMDriver/GCMDriver.jl [hunk 4]
```diff
@@ -217,25 +218,45 @@ function config_diagnostics(::Type{FT}, driver_config) where {FT}
...
 
     info = driver_config.config_info
+
+    # Setup diagnostic grid(s) 
+    nlats = 32
+    
+    sinθ, wts = ComputeGaussian!(nlats)
+    lats = asin.(sinθ) .* 180/π
+    lons = 180.0 ./ nlats *  collect(FT,1:1:2nlats)[:] .- 180.0
+    Δlons =  (lons[2]-lons[1])
```
Comment (jiahe23) [inline experiments/AtmosGCM/GCMDriver/GCMDriver.jl:228]: Δ is used as the Laplacian operator elsewhere and may cause some confusion.

File: src/Diagnostics/dump_spectra_gcm.jl [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    interpol = dgngrp.interpol
+    nor = dgngrp.params.nor
+
+    # get the 1d and 2d spectra and their associated wavenumbers
+    spectrum_1d, m, spectrum_2d, m_and_n = get_gcm_spectra(mpicomm, mpirank, Q, bl, interpol, nor)
+    if mpirank == 0
+        # Setup dimensions 
+        interpol_dims = dimensions(interpol)
+        lat = interpol_dims["lat"][1]
+        level = interpol_dims["level"][1]
```
Comment (jiahe23) [inline src/Diagnostics/dump_spectra_gcm.jl:94]: I think this is the distance from the Earth center. Earth radius should be removed to get the altitude heights.

Comment (kpamnany) [reply src/Diagnostics/dump_spectra_gcm.jl]: @LenkaNovak: I added a fix for this [here](https://github.com/CliMA/ClimateMachine.jl/pull/1559/files#diff-5a9fe607558ecfe8ea9b7ac212edf2a66e2b89d34d5d075a2e2b178428787d20R103-R107), please verify?

**Review by kpamnany — APPROVED** [review]

**Comment by kpamnany** [comment]
@LenkaNovak: I think this is good to go. Please take a look and if you're happy, we can merge it.

bors try

**Pull request merged**